flowapi/flowapi/api_spec.py

```diff
@@ -0,0 +1,121 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import yaml
+from apispec import APISpec, yaml_utils
+from quart import Blueprint, request, jsonify, render_template, current_app
+from zmq.asyncio import Socket
+from flowapi import __version__
+
+blueprint = Blueprint("spec", __name__)
+
+
+def remove_discriminators(spec: dict) -> dict:
+    """
+    Remove any discriminator keys from an openapi spec dict, because
+    they are not supported properly by redoc.
+
+    Parameters
+    ----------
+    spec : dict
+        Dict version of an openapi spec
+
+    Returns
+    -------
+    dict
+        The same spec, but with all discriminator keys removed.
+    """
+    newdict = {}
+    for k, v in spec.items():
+        if k != "discriminator":
+            newdict[k] = remove_discriminators(v) if isinstance(v, dict) else v
+    return newdict
+
+
+async def get_spec(socket: Socket, request_id: str) -> APISpec:
+    """
+    Construct open api spec by interrogating FlowMachine.
+
+    Parameters
+    ----------
+    socket : Socket
+    request_id : str
+        Unique id of the request
+
+    Returns
+    -------
+    APISpec
+        The specification object
+
+    """
+    msg = {"request_id": request_id, "action": "get_query_schemas"}
+    socket.send_json(msg)
+    #  Get the reply.
+    reply = await socket.recv_json()
+    flowmachine_query_schemas = reply["payload"]["query_schemas"]
+    # Need to mark query_kind as a required field
+    # this is a workaround because the marshmallow-oneOf plugin strips
```

Comment:
I get the feeling that there _should_ be some option to `marshmallow-oneofschema` to not strip it out, but it doesn't seem to support it at present. Might be worth raising an issue there.